components/automate-ui/src/app/page-components/simple-line-graph/simple-line-graph.component.ts

```diff
@@ -0,0 +1,342 @@
+  @Input() data: DayPercentage[] = [];
+  @Input() width = 900;
+  @Input() height = 156; // we want a 116px height on the ticks, this minus margins
+  private locked: number = null;
+
+  get margin() {
+    return { right: 20, left: 20, top: 20, bottom: 20 };
+  }
+  ////////   X AXIS ITEMS   ////////
+  // maps all of our x data points
+  get xData(): number[] {
+    return this.data.map(d => d.daysAgo);
+  }
+  // determines how wide the graph should be to hold our data
+  // in its respective area;
+  get rangeX(): [number, number] {
+    const min = this.margin.left;
+    const max = this.width - this.margin.right;
+    return [max, min];  // we want to plot our data backwards, so we reverse [min, max]
+  }
+  // determines the min and max values of the x axis
+  get domainX(): [number, number] {
+    const min = Math.min(...this.xData);
+    const max = Math.max(...this.xData);
+    return [min, max];
+  }
+
+  get xScale() {
+    return d3.scaleTime()
+      .range(this.rangeX)
+      .domain(this.domainX);
+  }
+
+
+  ////////   Y AXIS ITEMS   ////////
+  get rangeY(): [number, number] {
+    const min = this.margin.top;
+    const max = this.height - this.margin.bottom;
+    return [max, min];
+  }
+
+  get domainY(): [number, number] {
+    const min = 0;
+    const max = 100; // since this based on a percentage we are doing 0 to 100;
+    return [min, max];
+  }
+
+  get yScale() {
+    return d3.scaleLinear()
+             .range(this.rangeY)
+             .domain(this.domainY);
+  }
+
+  //////// SELECTIONS ////////
+  get containerSelection() {
+    return d3.select('app-simple-line-graph');
+  }
+
+  get svgSelection() {
+    return d3.select(this.svg.nativeElement);
+  }
+
+  get axisYSelection() {
+    return this.svgSelection.select('.y-axis');
+  }
+
+  // returns a function that when passed our data, will return an svg path
+  get createPath() {
+    return d3.line()
+              .x(d => this.xScale( d.daysAgo) )
+              .y(d => this.yScale( d.percentage) );
+  }
+
+  get viewBox() {
+    return `0 0 ${this.width} ${this.height}`;
+  }
+
+    ////////////////// RENDER FUNCTIONS ////////////////////
+  private renderChart() {
+    this.AllUnlockDeactivate();
+    this.resizeChart();
+    this.renderGrid();
+    this.renderLine();
+    this.renderPoints();
+    this.renderTooltips();
+    this.renderRings();
+    this.renderLabelButtons();
+    this.relock();
+  }
+
+  private renderLine(): void {
+    // create the line using path function
+    const line = this.createPath(this.data);
+
+    const theLine = this.svgSelection.selectAll('.line').data([this.data], d => d.daysAgo);
+    theLine.exit().remove();
+    theLine.enter().append('path').attr('class', 'line').merge(theLine)
+      .transition().duration(1000)
+    .attr('d', line);
+  }
+
+  private renderPoints(): void {
+    const points = this.svgSelection.selectAll('circle.point')
+      .data(this.data, d => d.daysAgo);
```

Comment:
you will notice that many of the d3 update patterns are passing a second parameter to `.data()`.  in this case i'm referring to ` d => d.daysAgo`.  This provides d3 a key to index the data on, which allows us to keep a reference when we're animating the data around.